pvanalytics/features/clipping.py

```diff
@@ -225,3 +225,178 @@ def threshold(ac_power, slope_max=0.0035, power_min=0.75,
         freq=freq
     )
     return ac_power >= threshold
+
+
+def _freq_minutes(index, freq):
+    """Return the frequency in minutes for `freq`. If `freq` is None
+    then use the frequency inferred from `index`."""
+    if freq is None:
+        freq = pd.infer_freq(index)
+    return util.freq_to_timedelta(freq).seconds / 60
+
+
+def _apply_daily_mask(mask, data, f):
+    """Apply `f` to the data selected by `mask` on each day.
+
+    Parameters
+    ----------
+    mask : Series
+        Boolean Series with same index as `data`
+    data : Series
+        Series with the data.
+    f : callable
+        Function that takes a series as an input and returns a
+        sing value.
+
+    Returns
+    -------
+    Series
+        Series with same index as `mask` and values returned
+        by applying `f` to ``data[mask]`` on each day.
+    """
+    daily = mask.resample('D')
+    return daily.transform(
+        lambda day: f(data[day.index][day])
+    )
+
+
+def _threshold_mean(mask, data):
+    """Return daily thresholds based on mean and standard deviation.
+
+    Parameters
+    ----------
+    mask : Series
+        Boolean series.
+    data : Series
+        Data with same index as `mask`.
+
+    Returns
+    -------
+    minimum : Series
+        `data` transformed to the mean of ``data[mask]`` minus 2 times
+         the standard deviation of ``data[mask]`` on each day.
+    maximum : Series
+        `data` transformed to the mean of ``data[mask]`` plus 2 times
+         the standard deviation of ``data[mask]`` on each day.
+    """
+    daily_mean = _apply_daily_mask(mask, data, pd.Series.mean)
```

Comment:
Curious why use the custom function here instead of something like `data[mask].rolling('D').mean()`. Because the index is uneven after `data[mask]`? 

As I understand it, using `lambda` or `apply` is a real performance hit, since those loop in python rather than applying some C code.

Reply:
Yeah, I did it this way so I could use `.transform()` and end up with the same index. The other way is doable, but ends up with a bunch of re-indexing at the end that I seemed error-prone and more difficult to maintain.

I'm not sure about the performance hit from using a `lambda`. I recall hearing the same in the past, but I have not found any recent references to performance loss, I think we would need to do some profiling to see if it really makes a difference here.

Reply:
OK as is

Reply:
If I'm the source of the claimed perf loss with `apply`, this is what I was complaining about:

```python
In [23]: import pandas as pd

In [24]: pd.__version__
Out[24]: '0.25.1'

In [25]: s = pd.Series(1, pd.date_range('2019-01-01', freq='h', periods=8760))

In [26]: %time _=s.rolling('d').apply(sum, raw=False)
Wall time: 906 ms

In [27]: %time _=s.rolling('d').apply(sum, raw=True)
Wall time: 36.9 ms

In [30]: %time _=s.rolling('d').sum()
Wall time: 998 µs
```

Reply:
Seems like that is down to the difference between applying the native python `sum()` and `pandas.core.window.Rolling.sum()`?

Reply:
I agree that comparison is not apples-to-apples, but I'm not immediately finding a way to use the Rolling methods with `apply`.  Do you have a demo snippet you can show?  If there's a way to use `apply` that has timings comparable to those of the rolling methods, I would love to know about it.

Reply:
Good point. You're right that you can't apply those directly in this context.

Reply:
I was thinking this would give the same results, but with improved performance:

```python
def _apply_daily_mask(mask, data, transformation):
    """Apply `f` to the data selected by `mask` on each day.

    Parameters
    ----------
    mask : Series
        Boolean Series with same index as `data`
    data : Series
        Series with the data.
    transformation : str or function
        Any value that can be passed to ``Series.resample().transform()``.

    Returns
    -------
    Series
        Series with same index as `mask` and values assigned by applying
        transformation to data in ``data[mask]`` on each day.
    """
    data = data.copy()
    data[~mask] = np.nan
    return data.resample('D').transform(transformation)
```

We can pass the names of cythonized kernel functions (`'min'`, `'max'`, `'mean'`, and `'std'`) as `transformation` and should get substantially better performance (5ms per call as opposed to 150ms per call). All well and good, but tests fail for 5-minute data.

Reply:
Using the cythonized functions gives numerically different results due to floating point rounding problems. For example, `tests/features/test_clipping.py::test_geometric_clipping_correct[5T]` ends up with an upper and a lower bound of `62.40000000000005`, while the maximum value in `ac_power` is `62.4` exactly.

Reply:
weird

Reply:
Can work around this by setting the minimum to be no larger than the absolute maximum candidate clipped value on each day. I'm not completely happy with this as it feels like a bit of a hack, but it does work for the tests I have implemented. It should work in general since the minimum will only ever be set on days where there is data that appears to be clipped. This also should not generally come up in practice; I believe it only arises here because the standard deviation is 0 since there is no noise in the simulated data used for testing. This is possible, but very unlikely in practice.

Reply:
Or maybe round the returned min, max values to some large number of decimal places, like 8? A comment is appropriate whatever you choose to do.

Reply:
I was starting to think rounding might be the better option. Still end up with one test failure, so I will need to dig in to it a bit more.